runtime/src/accounts_db.rs

```diff
@@ -12275,4 +12275,74 @@ pub mod tests {
         accounts_db.clean_accounts(None, false, Some(slot3));
         assert_eq!(accounts_db.ref_count_for_pubkey(&pubkey), 0);
     }
+
+    #[test]
+    fn test_filter_zero_lamport_clean_for_incremental_snapshots() {
+        solana_logger::setup();
+        let accounts_db = AccountsDb::new_single_for_tests();
+        let pubkey = solana_sdk::pubkey::new_rand();
+        let slot = 10;
+        let account_info = AccountInfo {
+            store_id: 42,
+            offset: 123,
+            stored_size: 234,
+            lamports: 0,
+        };
+        let store_count = 0;
+        let mut key_set = HashSet::default();
+        key_set.insert(pubkey);
+        let mut store_counts = HashMap::default();
+        store_counts.insert(account_info.store_id, (store_count, key_set));
+
+        // scenario 1: zero lamport account GREATER THAN last full snapshot slot => remove from purges set
+        {
+            let max_clean_root = None;
+            let last_full_snapshot_slot = Some(slot - 1);
+            let mut purges_zero_lamports = HashMap::default();
+            purges_zero_lamports.insert(pubkey, (vec![(slot, account_info.clone())], 1));
+
+            accounts_db.filter_zero_lamport_clean_for_incremental_snapshots(
+                max_clean_root,
+                last_full_snapshot_slot,
+                &store_counts,
+                &mut purges_zero_lamports,
+            );
+
+            assert!(!purges_zero_lamports.contains_key(&pubkey));
+        }
+
+        // scenario 2: zero lamport account EQUAL TO last full snapshot slot => keep in purges set
+        {
+            let max_clean_root = None;
+            let last_full_snapshot_slot = Some(slot);
+            let mut purges_zero_lamports = HashMap::default();
+            purges_zero_lamports.insert(pubkey, (vec![(slot, account_info.clone())], 1));
+
+            accounts_db.filter_zero_lamport_clean_for_incremental_snapshots(
+                max_clean_root,
+                last_full_snapshot_slot,
+                &store_counts,
+                &mut purges_zero_lamports,
+            );
+
+            assert!(purges_zero_lamports.contains_key(&pubkey));
+        }
+
+        // scenario 3: zero lamport account LESS THAN full snapshot slot => keep in purges set
+        {
+            let max_clean_root = None;
+            let last_full_snapshot_slot = Some(slot + 1);
+            let mut purges_zero_lamports = HashMap::default();
+            purges_zero_lamports.insert(pubkey, (vec![(slot, account_info)], 1));
+
+            accounts_db.filter_zero_lamport_clean_for_incremental_snapshots(
+                max_clean_root,
+                last_full_snapshot_slot,
+                &store_counts,
+                &mut purges_zero_lamports,
+            );
+
+            assert!(purges_zero_lamports.contains_key(&pubkey));
+        }
```

Comment:
These three cases would be nice:

1) `max_clean_root` and `last_full_snapshot_slot` both `None`
2) `max_clean_root` and `last_full_snapshot_slot` both `Some`
3) `max_clean_root` `Some` and `last_full_snapshot_slot` `None`.

Reply:
Done! Should cover a lot more cases now.